components/org.wso2.carbon.identity.local.auth.push.authenticator/src/main/java/org/wso2/carbon/identity/local/auth/push/authenticator/PushAuthenticator.java

```diff
@@ -0,0 +1,1702 @@
+                    request, context);
+            String pushDeviceRegistrationPageUrl = getPushDeviceEnrollConsentPageUrl();
+            String url = FrameworkUtils.appendQueryParamsStringToUrl(pushDeviceRegistrationPageUrl,
+                    queryParamsBuilder.toString());
+            response.sendRedirect(url);
+        } catch (Exception e) {
+            throw handleAuthErrorScenario(ERROR_CODE_ERROR_REDIRECTING_TO_DEVICE_REGISTRATION_PAGE, e, (Object) null);
+        }
+    }
+
+    /**
+     * Build query params for the IDF user device enrollment consent page.
+     *
+     * @param authenticatedUser Authenticated user.
+     * @param request           HttpServletRequest.
+     * @param context           AuthenticationContext.
+     * @return Query params for the IDF user device enrollment consent page.
+     */
+    protected StringBuilder buildQueryParamsForIDFUserDeviceEnrolConsentPage(AuthenticatedUser authenticatedUser,
+                                                                            HttpServletRequest request,
+                                                                            AuthenticationContext context) {
+
+        StringBuilder queryParamsBuilder = new StringBuilder();
+        String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                context.getCallerSessionKey(), context.getContextIdentifier());
+        String multiOptionURI = AuthenticatorUtils.getMultiOptionURIQueryString(request);
+        queryParamsBuilder.append(queryParams)
+                .append(AUTHENTICATORS_QUERY_PARAM).append(getName())
+                .append(USERNAME_PARAM).append(authenticatedUser.getUserName())
+                .append(multiOptionURI);
+        return queryParamsBuilder;
+    }
+
+    /**
+     * Retrieve the provisioned username of the authenticated user. If this is a federated scenario, the
+     * authenticated username will be same as the username in context. If the flow is for a JIT provisioned user, the
+     * provisioned username will be returned.
+     *
+     * @param authenticatedUser AuthenticatedUser.
+     * @param context           AuthenticationContext.
+     * @return Provisioned username
+     * @throws AuthenticationFailedException If an error occurred while getting the provisioned username.
+     */
+    private String getMappedLocalUsername(AuthenticatedUser authenticatedUser, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        if (!authenticatedUser.isFederatedUser()) {
+            return authenticatedUser.getUserName();
+        }
+        // If the user is federated, we need to check whether the user is already provisioned to the organization.
+        String federatedUsername = FederatedAuthenticatorUtil.getLoggedInFederatedUser(context);
+        if (StringUtils.isBlank(federatedUsername)) {
+            throw handleAuthErrorScenario(ERROR_CODE_NO_FEDERATED_USER);
```

Comment:
Check the need for this exception. The usages of this method (getMappedLocalUsername) accepts an empty value.